File: scripts/javascript/token-contract/supply.js
const hre = require("hardhat");

async function main() {
  // Replace with your deployed factory address
  const factoryAddr = process.env.FACTORY_ADDRESS;
  if (!factoryAddr) {
    throw new Error(
      "FACTORY_ADDRESS is not set."
    );
  }

  const factory = await hre.ethers.getContractAt("TokenFactory", factoryAddr);
  const tokens = await factory.getAllTokens();

  const tokenAddr = tokens[tokens.length - 1];
  console.log("Using token:", tokenAddr);

  const token = await hre.ethers.getContractAt("CustomToken", tokenAddr);

  // Call totalSupply()
  const supply = await token.totalSupply();
  console.log("Total Supply:", supply);
}

main()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });

File: scripts/rust/token-contract/src/bin/supply.rs
use alloy::{primitives::Address, providers::ProviderBuilder, signers::local::PrivateKeySigner};
use std::{env, str::FromStr};
use token_contract::CustomToken;
use url::Url;

#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
    dotenv::dotenv().ok();

    let contract_str = std::env::var("TOKEN_ADDRESS")?;
    let contract = Address::from_str(&contract_str)?;
    let private_key = std::env::var("PRIVATE_KEY")?;

    // Alloy provider + signer
    let signer: PrivateKeySigner = PrivateKeySigner::from_str(&private_key)?;

    // Set up the WS transport and connect.
    let rpc_url = env::var("RPC_URL")?;
    let provider = ProviderBuilder::new()
        .wallet(signer)
        .connect_http(Url::parse(&rpc_url)?);

    let factory = CustomToken::new(contract, provider);
    let supply = factory.totalSupply().call().await?;
    println!("✅ Total Supply: {supply:?}");

    Ok(())
}
